src/aes/mod.rs

```diff
@@ -1 +1,116 @@
 pub mod registers;
+
+use crate::error::{ErrorKind, Result};
+use crate::memory_map::mmio;
+use registers::Registers;
+
+pub struct AES {
+    registers: Registers,
+}
+
+pub enum Key<'a> {
+    Bits128(&'a [u8; 16]),
+    Bits192(&'a [u8; 24]),
+    Bits256(&'a [u8; 32]),
+}
+
+const ENCRYPT_EXTERNAL_KEY: u8 = 0b_00;
+const DECRYPT_EXTERNAL_KEY: u8 = 0b_01;
+// FIXME User Guide is vague and local generation doesn't make sense
+//const DECRYPT_LOCAL_KEY: u8 = 0b_10;
+
+impl AES {
+    pub fn init() -> Self {
+        Self {
+            registers: Registers::new(mmio::AES),
+        }
+    }
+
+    pub fn encrypt(&mut self, data: &mut [u8], key_size: Key) -> Result<()> {
+        self.external_key(data, key_size, false)
+    }
+
+    pub fn decrypt(&mut self, data: &mut [u8], key_size: Key) -> Result<()> {
+        self.external_key(data, key_size, true)
+    }
+
+    fn external_key(&mut self, data: &mut [u8], key_size: Key, decrypt: bool) -> Result<()> {
+        if data.len() % 16 != 0 {
+            return Err(ErrorKind::BadParam);
+        }
+
+        // See MAX78000 User Guide 24.2 Encryption of 128-Bit Blocks of Data Using FIFO
+
+        let (key_ptr, key_len) = match key_size {
+            Key::Bits128(key) => (key.as_ref() as *const [u8] as *const u8, 16),
+            Key::Bits192(key) => (key.as_ref() as *const [u8] as *const u8, 24),
+            Key::Bits256(key) => (key.as_ref() as *const [u8] as *const u8, 32),
+        };
```

Comment:
Last time I think I forgot about `as_mut_ptr()` [see here](https://doc.rust-lang.org/std/primitive.slice.html#method.as_mut_ptr). I think it should be wise to switch this to that, or to use `address_of!`.

Reply:
Fake news, use `as_ptr()`. Remove `mut`.